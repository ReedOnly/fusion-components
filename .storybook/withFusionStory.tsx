import React from 'react';
import Header from '../src/components/core/Header';
import Content from '../src/components/core/Content';
import Overlay from '../src/components/general/ApplicationGuidance/components/Overlay';

export default (title: string, padding: number = 16) => (stories: () => React.ReactNode) => {
    return (
        <>
            <Overlay fixed>
                <Header aside={null} start={null} content={null} />
            </Overlay>
            <Content>
                <Overlay>
                    <div
                        style={{
                            padding,
                            minHeight: '100%',
                            width: '100%',
                            maxWidth: '100%',
                            boxSizing: 'border-box',
                        }}
                    >
                        {stories()}
                    </div>
                </Overlay>
            </Content>
        </>
    );
};
