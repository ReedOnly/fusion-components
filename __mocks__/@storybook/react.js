import renderer from 'react-test-renderer';
import addons, { mockChannel } from '@storybook/addons';
import { withFusionContext } from '../../.storybook/withFusionContext';

addons.setChannel(mockChannel());

export const action = (actionName) => jest.fn();
const fusionContext = withFusionContext();
export const storiesOf = (groupName) => {
    const api = {
        decorators: [],
        parameters: {},

        add(storyName, story) {
            if (api.parameters.skipTest) {
                return;
            }

            describe(groupName, () => {
                it(storyName, () => {
                    // const decoratedStory = api.decorators.reduce((s, decorator) => {
                    //     try {
                    //         return decorator(s, { parameters: api.parameters });
                    //     } catch (e) {
                    //         return s;
                    //     }
                    // }, story);

                    const component = renderer.create(fusionContext(story));

                    // @TODO Enable custom names once released:
                    // > https://github.com/facebook/jest/pull/2094
                    expect(component.toJSON())
                        .toMatchSnapshot
                        // `${groupName}.${storyName}`
                        ();
                });
            });

            return api;
        },

        addDecorator() {
            return api;
        },

        addParameters(parameters) {
            api.parameters = { ...api.parameters, ...parameters };
            return api;
        },
    };

    return api;
};

export const addDecorator = () => {};
