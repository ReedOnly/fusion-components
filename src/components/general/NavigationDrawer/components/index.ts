export { default as MenuChild } from "./Child";