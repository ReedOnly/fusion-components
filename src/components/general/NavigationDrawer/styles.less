@import '~styles/mixins.less';

.container {
    box-sizing: border-box;
    height: 100%;
    border-right: 2px solid var(--color-black-alt4);
    transition: width .1s;
    overflow-y: auto;
    overflow-x: hidden;
    color: var(--color-primary-accent);
    width: .grid-unit(40px)[];
    flex-shrink: 0;

    .collapseButtonContainer {
        width: .grid-unit(6px)[];
        height: .grid-unit(6px)[];
        border-bottom: 1px solid var(--color-black-alt4);
    }

    &.isCollapsed {
        width: .grid-unit(6px)[];
    }
}