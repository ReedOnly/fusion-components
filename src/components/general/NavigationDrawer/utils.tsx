import React from 'react';
import {
    NavigationChild,
    NavigationSection,
    NavigationGrouping,
    NavigationLabel,
} from './components';
import { NavigationStructure } from '.';

export const getNavigationComponentForItem = (structure?: NavigationStructure[], customProps?) => {
    if (!structure) {
        return null;
    }
    return structure.map(item => {
        const itemProps = {
            title: item.title,
            id: item.id,
            key:item.id,
            onClick: item.onClick,
            isActive: item.isActive,
            isOpen: item.isOpen,
            icon: item.icon,
            navigationChildren: item.navigationChildren,
        };
        switch (item.type) {
            case 'child':
                return <NavigationChild {...itemProps} {...customProps} />;
            case 'section':
                return <NavigationSection {...itemProps} {...customProps} />;
            case 'grouping':
                return <NavigationGrouping {...itemProps} {...customProps} />;
            case 'label':
                return <NavigationLabel {...itemProps} {...customProps} />;
            default:
                return null;
        }
    });
};

export const toggleOpenById = (id: string, structure: NavigationStructure): NavigationStructure => {
    if (structure.id === id) {
        return {
            ...structure,
            isOpen: !structure.isOpen,
        };
    }
    if (structure.navigationChildren) {
        return {
            ...structure,
            navigationChildren: structure.navigationChildren.map(child =>
                toggleOpenById(id, child)
            ),
        };
    }
    return structure;
};

export const toggleActiveById = (
    id: string,
    structure: NavigationStructure
): NavigationStructure => {
    if (structure.navigationChildren) {
        return {
            ...structure,
            isActive: structure.id === id,
            navigationChildren: structure.navigationChildren.map(child =>
                toggleActiveById(id, child)
            ),
        };
    }
    return {
        ...structure,
        isActive: structure.id === id,
    };
};
