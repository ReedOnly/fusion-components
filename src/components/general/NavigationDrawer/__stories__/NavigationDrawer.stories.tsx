import React from "react";
import { storiesOf } from '@storybook/react';
import withFusionStory from "../../../../../.storybook/withFusionStory";
import NavigationDrawer from "../index";

const NavigationDrawerStory = () => {
    return(
        <>
            <NavigationDrawer id="navigation-drawer-story" />
        </>
    )
}

storiesOf('General|Navigation Drawer', module)
    .addDecorator(withFusionStory('Navigation Drawer'))
    .add("Default", () => <NavigationDrawerStory />)



const test = [
    {
        key:"group1",
        type:"grouping",
        icon:() => <div>Icon</div>,
        title:"Menu Grouping",
        children:[
            {
                key:"section1",
                type:"section",
                title:"Menu Section",
                children: [
                    {
                        key:"child",
                        type:"Child",
                        title:"Menu child",
                        isSelected: true,
                    },
                    {
                        key:"child1",
                        type:"Child",
                        title:"Menu child1",
                    }
                ]
            }
        ]
    }
]