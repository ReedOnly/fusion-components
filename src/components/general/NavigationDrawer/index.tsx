import React, { FC, useState, useCallback, ReactNode, useMemo } from 'react';
import classNames from 'classnames';
import { useComponentDisplayClassNames } from '@equinor/fusion';
import { CollapseExpandButton } from '@equinor/fusion-components';
import styles from './styles.less';
import { NavigationChild, NavigationSection, NavigationGrouping } from './components';
import { getNavigationComponentForItem, toggleOpenById, toggleActiveById } from './utils';
export { NavigationChild, NavigationSection, NavigationGrouping };

const NAVIGATION_DRAWER_COLLAPSED_KEY = 'NAVIGATION_DRAWER_COLLAPSED_KEY';
const createDrawerCollapsedKey = (key: string) => NAVIGATION_DRAWER_COLLAPSED_KEY + key;
const getDefaultCollapsed = (key: string) => {
    const value = localStorage.getItem(createDrawerCollapsedKey(key));
    return Boolean(value);
};

const persistCollapsedState = (key: string, isCollapsed: boolean) => {
    localStorage.setItem(createDrawerCollapsedKey(key), isCollapsed ? 'collapsed' : '');
};

export type NavigationComponentProps = NavigationStructure & {
    onChange?: (id: string, toggleOpen: boolean, toggleActive: boolean) => void;
    isCollapsed?: boolean;
};

export type NavigationStructure = {
    id: string;
    type: 'child' | 'grouping' | 'label' | 'search' | 'section';
    icon?: ReactNode;
    title: string;
    onClick?: () => void;
    isActive?: boolean;
    isOpen?: boolean;
    navigationChildren?: NavigationStructure[];
};

type NavigationDrawerProps = {
    id: string;
    structure: NavigationStructure[];
    onChange: (newStructure: NavigationStructure[]) => void;
};

const NavigationDrawer: FC<NavigationDrawerProps> = ({ id, structure, onChange }) => {
    const [isCollapsed, setIsCollapsed] = useState(getDefaultCollapsed(id));

    const toggleCollapsed = useCallback(() => {
        persistCollapsedState(id, !isCollapsed);
        setIsCollapsed(!isCollapsed);
    }, [isCollapsed]);

    const containerClassNames = classNames(
        styles.container,
        useComponentDisplayClassNames(styles),
        {
            [styles.isCollapsed]: isCollapsed,
        }
    );
    const navigationStructure = useMemo(
        () =>
            getNavigationComponentForItem(structure, {
                onChange: (id: string, toggleOpen: boolean, toggleActive: boolean) => {
                    const newStructure = structure.map(item => {
                        if (toggleOpen && toggleActive) {
                            return toggleActiveById(id, toggleOpenById(id, item));
                        }
                        if (toggleOpen) {
                            return toggleOpenById(id, item);
                        }
                        if (toggleActive) {
                            return toggleActiveById(id, item);
                        }
                        return item;
                    });
                    newStructure && onChange(newStructure);
                },
                isCollapsed: isCollapsed,
            }),
        [structure, onChange, isCollapsed]
    );
    return (
        <div className={containerClassNames}>
            <div className={styles.collapseButtonContainer}>
                <CollapseExpandButton isCollapsed={isCollapsed} onClick={toggleCollapsed} />
            </div>
            {navigationStructure}
        </div>
    );
};

export default NavigationDrawer;
