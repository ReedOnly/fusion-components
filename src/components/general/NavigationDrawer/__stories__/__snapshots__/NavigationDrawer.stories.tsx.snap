// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General|Navigation Drawer Default 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      className="container comfortable"
    >
      <div
        className="collapseButtonContainer"
      >
        <button
          className="container comfortable"
          onClick={[Function]}
          onMouseDown={[Function]}
        >
          <span
            className="iconContainer"
          >
            <svg
              height={24}
              style={
                Object {
                  "color": "",
                }
              }
              viewBox="0 0 24 24"
              width={24}
            >
              <path
                clipRule="evenodd"
                d="M7.99878 7C7.99878 6.45 7.54878 6 6.99878 6C6.44878 6 5.99878 6.45 5.99878 7V17C5.99878 17.55 6.44878 18 6.99878 18C7.54878 18 7.99878 17.55 7.99878 17V7ZM13.8188 12L17.6988 15.89C18.0888 16.27 18.0888 16.91 17.7088 17.29C17.3188 17.68 16.6888 17.68 16.2988 17.29L11.7088 12.7C11.3188 12.31 11.3188 11.68 11.7088 11.29L16.2988 6.7C16.4856 6.51275 16.7393 6.40751 17.0038 6.40751C17.2683 6.40751 17.5219 6.51275 17.7088 6.7C18.0988 7.09 18.0988 7.72 17.7088 8.11L13.8188 12Z"
                fill="currentColor"
                fillRule="evenodd"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
